OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,454 @@
+    VkResult result = dispatch_table_.ResetCommandPool(device)(device, command_pool, flags);
+    submission_tracker_.ResetCommandPool(command_pool);
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnAllocateCommandBuffers(VkDevice device,
+                                                  const VkCommandBufferAllocateInfo* allocate_info,
+                                                  VkCommandBuffer* command_buffers) {
+    VkResult result =
+        dispatch_table_.AllocateCommandBuffers(device)(device, allocate_info, command_buffers);
+
+    VkCommandPool pool = allocate_info->commandPool;
+    const uint32_t command_buffer_count = allocate_info->commandBufferCount;
+    submission_tracker_.TrackCommandBuffers(device, pool, command_buffers, command_buffer_count);
+    return result;
+  }
+
+  void OnFreeCommandBuffers(VkDevice device, VkCommandPool command_pool,
+                            uint32_t command_buffer_count, const VkCommandBuffer* command_buffers) {
+    submission_tracker_.UntrackCommandBuffers(device, command_pool, command_buffers,
+                                              command_buffer_count);
+    return dispatch_table_.FreeCommandBuffers(device)(device, command_pool, command_buffer_count,
+                                                      command_buffers);
+  }
+
+  [[nodiscard]] VkResult OnBeginCommandBuffer(VkCommandBuffer command_buffer,
+                                              const VkCommandBufferBeginInfo* begin_info) {
+    VkResult result =
+        dispatch_table_.BeginCommandBuffer(command_buffer)(command_buffer, begin_info);
+    submission_tracker_.MarkCommandBufferBegin(command_buffer);
+    return result;
+  }
+
+  [[nodiscard]] VkResult OnEndCommandBuffer(VkCommandBuffer command_buffer) {
+    submission_tracker_.MarkCommandBufferEnd(command_buffer);
+    return dispatch_table_.EndCommandBuffer(command_buffer)(command_buffer);
+  }
+
+  [[nodiscard]] VkResult OnResetCommandBuffer(VkCommandBuffer command_buffer,
+                                              VkCommandBufferResetFlags flags) {
+    submission_tracker_.ResetCommandBuffer(command_buffer);
+    return dispatch_table_.ResetCommandBuffer(command_buffer)(command_buffer, flags);
+  }
+
+  void OnGetDeviceQueue(VkDevice device, uint32_t queue_family_index, uint32_t queue_index,
+                        VkQueue* queue) {
+    dispatch_table_.GetDeviceQueue(device)(device, queue_family_index, queue_index, queue);
+    queue_manager_.TrackQueue(*queue, device);
+  }
+
+  void OnGetDeviceQueue2(VkDevice device, const VkDeviceQueueInfo2* queue_info, VkQueue* queue) {
+    dispatch_table_.GetDeviceQueue2(device)(device, queue_info, queue);
+    queue_manager_.TrackQueue(*queue, device);
+  }
+
+  [[nodiscard]] VkResult OnQueueSubmit(VkQueue queue, uint32_t submit_count,
+                                       const VkSubmitInfo* submits, VkFence fence) {
+    auto queue_submission_optional =
```

Comment:
Could you state the type?

Reply:
Don't you have all the types available here? Of course they use the type parameters.

Reply:
No, this is an optional<SubmissionTracker::Submission>. Those types are not there.

Reply:
Ah, the return type is different in the tests, I noticed `// Note we simplify the return type, as it is not really part of this test.`. Why not add an empty `struct QueueSubmission{}` to `MockSubmissionTracker`? It seems better than using `bool`.

Then you should (maybe...) be able to write 
`std::optional<typename SubmissionTracker::QueueSubmission> queue_submission_optional =`.